src/web.rs

```diff
@@ -181,8 +225,10 @@ impl WebImpl {
             self.canvas
                 .put_image_data(
                     &image_data,
-                    rect.x.into(),
-                    rect.y.into(),
+                    union_region.left.into(),
+                    union_region.top.into(),
+                    (rect.x - union_region.left).into(),
+                    (rect.y - union_region.top).into(),
```

Comment:
> When the actual `ImageData` is as big as the canvas itself, `dx` and `dy` should always be `0.0` and `0.0`, and only `dirtyX` and `dirtyY` are actually representing the top-left coordinates for the dirty region. The dirty region is relative to the image data.

This is what I’m referring to here. Without the optimization we have `union_region.left = 0` and `union_region.top = 0`, resulting in something like:
```rust
                    0.0,
                    0.0,
                    rect.x.into(),
                    rect.y.into(),
```

Reply:
I'm not following the logic here, why do we pass `union_region.left` to `dx`?

As far as I understand `dx` is where to draw relative to the destination image, `union_region.left` is the X coordinate of our coalesced dirty `Rect`s compared to the destination image.

Reply:
I assume it's what you were trying to explain in OP, but I don't seem to really get it. As far as I can tell the original implementation worked correctly.

Reply:
I’m sorry, my explanation was likely not clear enough.

> As far as I understand `dx` is where to draw relative to the destination image

Yes, to be pedantic, it is where to draw the _image data_ on the _destination canvas_.

The original implementation is [building a bitmap / image data for the _whole canvas_](https://github.com/rust-windowing/softbuffer/blob/9b44f2eaea136cae30543ecbad96827af7ff579c/src/web.rs#L144-L150).
As such, `dx` (respectively `dy`) was supposed to be `0.0`, because in that case the top left pixel of the image data is matching the top left pixel in the destination canvas.

However, the original implementation is [not passing `0.0`](https://github.com/rust-windowing/softbuffer/blob/9b44f2eaea136cae30543ecbad96827af7ff579c/src/web.rs#L182-L188), it passes [`rect.x` (respectively `rect.y`)](https://github.com/rust-windowing/softbuffer/blob/9b44f2eaea136cae30543ecbad96827af7ff579c/src/web.rs#L285-L293).

Next, `dirtyX` is the X-cordinate of the top-left corner of the damaged region _inside the image data_, NOT in the destination canvas.

To clarify: this does work correctly when the damage rectangle have _exactly_ the same dimension as the canvas, i.e.: `present` is not broken for this reason because [it happens to use the only valid damage rectangle](https://github.com/rust-windowing/softbuffer/blob/9b44f2eaea136cae30543ecbad96827af7ff579c/src/web.rs#L325-L330) for which the computation gives the correct result.

Here is an example. Let’s assume we have a 10x10 pixels canvas and let `data` be the image data for the whole canvas (again, 10x10) and `dirty = { top: 5, left: 5, bottom: 8, right: 8 }` be the damaged region.

Given the above points, the correct call is:
```js
canvas.putImageData(data, 0, 0, 5, 5, 3, 3);
```
Which roughly means "write the 10x10 image data at (0, 0) in the canvas with a dirty region from (5, 5) to (8, 8) relative to the image data". In this case only the pixels from (5, 5) to (8, 8) (the dirty region) are extracted from the image data and actually drawn.

The original implementation is making this call:
```js
canvas.putImageData(data, 5, 5, 5, 5, 3, 3);
```

I hope it helps clarifying a bit.

Reply:
> why do we pass `union_region.left` to `dx`?

Building upon the previous answer: in my patch, the image data holds the pixels for the union of all the damaged regions only, and as such does not necessarily covers the left-most part of the canvas as the original implementation does.

Reply:
I honestly can't interpret it this way from the MDN documentation, but the WHATWG specification was much clearer to me and seems to be in agreement with what you are saying. I also tested this locally and can confirm that this is correct.

Thanks!

Reply:
To be fair, it tookme some time playing around with the visual artifacts to finally reach this conclusion; the MDN doc could be clearer